src/drivers/distance_sensor/leddar_one/leddar_one.cpp

```diff
@@ -228,104 +216,86 @@ LeddarOne::crc16_calc(const unsigned char *data_frame, uint8_t crc16_length)
 int
 LeddarOne::collect()
 {
-	if (!_collect_phase) {
-		return measure();
-	}
-
-	perf_end(_sample_perf);
-
-	const hrt_abstime time_now = hrt_absolute_time();
+	perf_begin(_sample_perf);
 
-	struct reading_msg *msg;
+	const int buffer_size = sizeof(_buffer);
+	const int message_size = sizeof(reading_msg);
 
-	int bytes_read = ::read(_file_descriptor, _buffer + _buffer_len, sizeof(_buffer) - _buffer_len);
+	int bytes_read = ::read(_file_descriptor, _buffer + _buffer_len, buffer_size - _buffer_len);
 
 	if (bytes_read < 1) {
-		if (time_now > _timeout_usec) {
-			_collect_phase = true;
-			_timeout_usec = 0;
-		}
-
-		perf_count(_collect_timeout_perf);
-		return PX4_ERROR;
+		// Trigger a new measurement.
+		return measure();
 	}
 
 	_buffer_len += bytes_read;
 
-	if (_buffer_len < sizeof(struct reading_msg)) {
-		perf_count(_comms_errors);
-		return PX4_ERROR;
+	if (_buffer_len < message_size) {
+		// Return on next scheduled cycle to collect remaining data.
+		return PX4_OK;
 	}
 
-	msg = (struct reading_msg *)_buffer;
+	reading_msg *msg {nullptr};
+	msg = (reading_msg *)_buffer;
 
-	if (msg->slave_addr != MODBUS_SLAVE_ADDRESS || msg->function != MODBUS_READING_FUNCTION) {
-		perf_count(_comms_errors);
-		return PX4_ERROR;
+	if (msg->slave_addr != MODBUS_SLAVE_ADDRESS ||
+	    msg->function != MODBUS_READING_FUNCTION) {
+		PX4_ERR("slave address or function read error");
+		perf_count(_comms_error);
+		perf_end(_sample_perf);
+		return measure();
 	}
 
-	const uint16_t crc16 = crc16_calc(_buffer, _buffer_len - 2);
+	const uint16_t crc16 = crc16_calc(_buffer, buffer_size - 2);
```

Comment:
So this always does the CRC over the whole buffer (-2) and not just what has been read?

Reply:
Hi @julianoes , yes, that's correct, but the method will never get this far unless the buffer was filled.